### PR TITLE
Translate microscopic cross section lookups to C++

### PR DESCRIPTION
**Pull request by paulromano**
This PR moves the `Nuclide % calculate_xs` type-bound procedure over to C++ along with a number of related functions. Everything related to windowed multiple is now part of `wmp.cpp`. That helped simplify the usage of the Faddeeva library (namely, we now compile it using C++ rather than C because we don't need C bindings). Thermal scattering data is now stored in the `openmc::data::thermal_scatt` vector.

There is definitely some duplicated code/data between `nuclide_header.F90` and `nuclide.cpp` right now, but I'm hoping to whittle down the F90 side some more in the near future.

### CODE REVIEW

File: src/wmp.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  // Bookkeeping
+
+  // Define some frequently used variables.
+  double sqrtE = std::sqrt(E);
+  double invE = 1.0 / E;
+  double T = sqrtkT*sqrtkT / K_BOLTZMANN;
+
+  if (sqrtkT == 0.0) {
+    fatal_error("Windowed multipole temperature derivatives are not implemented"
+      "for 0 Kelvin cross sections.");
```
Comment (smharper) [inline src/wmp.cpp]: Need whitespace here, `" for"`

File: include/openmc/wmp.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+constexpr int FIT_F {2}; // Fission
+
+//========================================================================
+// Windowed multipole data
+//========================================================================
+
+class WindowedMultipole {
+public:
+  // Types, aliases
+  using cdouble = std::complex<double>;
```
Comment (smharper) [inline include/openmc/wmp.h]: This alias is only used in one place so I think we may as well just stick with the long name.

File: include/openmc/wmp.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  // Constructors, destructors
+  WindowedMultipole(hid_t group);
+
+  // Methods
+
+  //! Evaluate the windowed multipole equations for cross sections in the
+  //! resolved resonance regions
+  //!
```
Comment (smharper) [inline include/openmc/wmp.h]: Doxygen is too picky to use that as the "brief" method description.  It needs to be on one line and separated from the rest with `//` instead of `//!`.  If you can't make it fit on one line, use the `\brief` statement.

File: vendor/faddeeva/Faddeeva.hh [hunk 4]
```diff
@@ -1,68 +1,62 @@
...
- *
+ * 
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
  * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
  * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
  * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
- * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (smharper) [inline vendor/faddeeva/Faddeeva.hh]: Remove the whitespace added here.

File: vendor/faddeeva/Faddeeva.hh [hunk 5]
```diff
@@ -1,68 +1,62 @@
...
 
 /* Available at: http://ab-initio.mit.edu/Faddeeva
 
-   Header file for Faddeeva.c; see Faddeeva.cc for more information. */
+   Header file for Faddeeva.cc; see that file for more information. */
 
-#ifndef FADDEEVA_H
-#define FADDEEVA_H 1
+#ifndef FADDEEVA_HH
+#define FADDEEVA_HH 1
```
Comment (smharper) [inline vendor/faddeeva/Faddeeva.hh]: Why did you change `.h` to `.hh`?

File: src/nuclide.cpp [hunk 6]
```diff
@@ -404,7 +497,245 @@ double Nuclide::elastic_xs_0K(double E) const
...
+    // Find the appropriate temperature index.
+    double kT = sqrtkT*sqrtkT;
+    double f;
+    switch (settings::temperature_method) {
+    case TEMPERATURE_NEAREST:
+      {
+        double max_diff = INFTY;
+        for (int t = 0; t < kTs_.size(); ++t) {
+          if (std::abs(kTs_[t] - kT) < max_diff) {
+            i_temp = t;
```
Comment (smharper) [inline src/nuclide.cpp]: Am I reading this correctly?  It looks like this will always set `i_temp = 0`

File: src/nuclide.cpp [hunk 7]
```diff
@@ -404,7 +497,245 @@ double Nuclide::elastic_xs_0K(double E) const
...
+      (grid.energy[i_grid + 1]- grid.energy[i_grid]);
+
+    micro_xs.index_temp = i_temp;
+    // TODO: off-by-one
+    micro_xs.index_grid = i_grid + 1;
+    micro_xs.interp_factor = f;
+
+    // Calculate microscopic nuclide total cross section
+    micro_xs.total = (1.0 - f)*xs(i_grid,XS_TOTAL)
+          + f*xs(i_grid + 1,XS_TOTAL);
```
Comment (smharper) [inline src/nuclide.cpp]: Add commas in these `xs()` calls for readability.

File: src/nuclide.cpp [hunk 8]
```diff
@@ -404,7 +497,245 @@ double Nuclide::elastic_xs_0K(double E) const
...
+            // the threshold for (n,3n) and (n,4n).
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  // Initialize sab treatment to false
+  micro_xs.index_sab = -1;
```
Comment (smharper) [inline src/nuclide.cpp]: Should we use `C_NONE` here?

File: src/nuclide.cpp [hunk 9]
```diff
@@ -404,7 +497,245 @@ double Nuclide::elastic_xs_0K(double E) const
...
+  // sections.
+
+  if (i_sab >= 0) this->calculate_sab_xs(i_sab, E, sqrtkT, sab_frac);
+
+  // If the particle is in the unresolved resonance range and there are
+  // probability tables, we need to determine cross sections from the table
+  if (settings::urr_ptables_on && urr_present_ && !use_mp) {
+    int n = urr_data_[i_temp].n_energy_;
+    if ((E > urr_data_[i_temp].energy_(0)) &&
+        (E < urr_data_[i_temp].energy_(n-1))) {
```
Comment (smharper) [inline src/nuclide.cpp]: Can we use `micro_xs.index_temp` here instead of `i_temp` to avoid that awkward declaration of `i_temp` outside the `if` block?

File: src/nuclide.cpp (hunk 6 above)
Comment (paulromano) [reply src/nuclide.cpp]: You're right -- there's something missing here. I'll fix it. Thanks!

File: vendor/faddeeva/Faddeeva.hh (hunk 4 above)
Comment (smharper) [reply vendor/faddeeva/Faddeeva.hh]: Ah, didn't realize there were two different headers available.  Agreed, let's keep everything untouched.

**Comment by smharper** [comment]
Everything here looks good.  I also tried a calc with MCNP cross sections and temperature interpolation and I didn't find a regression.  Thanks for the PR, @paulromano!
